### PR TITLE
docs: mutable pitfalls

### PR DESCRIPTION
**Pull request by tthsqe12**

### CODE REVIEW

File: docs/src/DeveloperDocumentation/new_developers.md [hunk 1]
```diff
@@ -295,3 +295,133 @@ git fetch oscar-system
 git rebase oscar-system/master
 ```
 Adding the remote only has to be executed once.
+
+### Pitfalls: Mutable objects in Oscar code
+
+The objects in Oscar that can be changed (mutable objects) can be linked
```
Comment (fingolfin) [inline docs/src/DeveloperDocumentation/new_developers.md]: But this isn't Oscar specific, isn't it? This is affects all of Julia (or for that matter, Python, Ruby, Java, GAP, ...)

File: docs/src/DeveloperDocumentation/new_developers.md [hunk 2]
```diff
@@ -295,3 +295,133 @@ git fetch oscar-system
...
+
+```julia
+julia> a = Hecke.modular_proj(1+2*i, m)
+2-element Vector{fq_nmod}:
+ 2
+ 0
+```
+
+While the function has produced the correct answer, if we run it again on a
+different input, we will find that `a` has changed.
```
Comment (fingolfin) [inline docs/src/DeveloperDocumentation/new_developers.md]: That seems like a genuine bug in `modular_proj` to me? (Without having looked at its documentation or knowing anything about it at all...)

File: docs/src/DeveloperDocumentation/new_developers.md [hunk 3]
```diff
@@ -295,3 +295,133 @@ git fetch oscar-system
...
+
+julia> a = one!(a)  # modyfy a
+1
+
+julia> evaluate(x)  # x is now unchanged
+8
+```
+
+The previous two examples dealt with the fact that ownership conventions on
+function *arguments* are completely unspecified in Oscar. If you are going to
```
Comment (fingolfin) [inline docs/src/DeveloperDocumentation/new_developers.md]: I don't find this style of writing very constructive or productive. You are right to point out an issue, but how about us working to rectify it by trying to formulate some rules how ownership for arguments and return values should work by default? Then we could document that, and start to fix functions that violate the rules (by changing their behavior, or perhaps in some cases, by changing their documentation to explicitly point out the deviation).

Also, once again, I am fairly certain this problem is not unique to Oscar within the Julia realm, or in general.

File: docs/src/DeveloperDocumentation/new_developers.md (hunk 2 above)
Comment (thofma) [reply docs/src/DeveloperDocumentation/new_developers.md]: I think this is fine. This tries to allocate as few things as needed.

**Comment by fieker** [comment]
On Mon, Sep 26, 2022 at 02:17:20PM -0700, Max Horn wrote:
> @fingolfin commented on this pull request.
> 
> 
> 
> > @@ -295,3 +295,133 @@ git fetch oscar-system
>  git rebase oscar-system/master
>  ```
>  Adding the remote only has to be executed once.
> +
> +### Pitfalls: Mutable objects in Oscar code
> +
> +The objects in Oscar that can be changed (mutable objects) can be linked
> 
> But this isn't Oscar specific, isn't it? This is affects all of Julia (or for that matter, Python, Ruby, Java, GAP, ...)
> 
> > +julia> m = Hecke.modular_init(K, 5)
> +modular environment for p=5, using 2 ideals
> +```
> +
> +The function `modular_project` returns the projection of an element of `K` into
> +each of the residue fields.
> +
> +```julia
> +julia> a = Hecke.modular_proj(1+2*i, m)
> +2-element Vector{fq_nmod}:
> + 2
> + 0
> +```
> +
> +While the function has produced the correct answer, if we run it again on a
> +different input, we will find that `a` has changed.
> 
> That seems like a genuine bug in `modular_proj` to me? (Without having looked at its documentation or knowing anything about it at all...)
> 
This is actually deliberate. modular_proj and the entire family are
there tp faciliate high-performance modular algorithms, thus take as
many shortcuts as possible
 - dealing only with "easy" "small" primes
   (no denominators, no ramification, no index divisors)
 - do not check anything
 - try to allocate as little as possible using inplace operations where
   feasible
They are not meant for the faint-hearted, the casual user or a beginner.
Maybe they need to be locked away under some expert only section.
> > +```julia
> +julia> a = fmpz(2)
> +2
> +
> +julia> x = FacElem([deepcopy(a)], [fmpz(3)]); evaluate(x)
> +8
> +
> +julia> a = one!(a)  # modyfy a
> +1
> +
> +julia> evaluate(x)  # x is now unchanged
> +8
> +```
> +
> +The previous two examples dealt with the fact that ownership conventions on
> +function *arguments* are completely unspecified in Oscar. If you are going to
> 
> I don't find this style of writing very constructive or productive. You are right to point out an issue, but how about us working to rectify it by trying to formulate some rules how ownership for arguments and return values should work by default? Then we could document that, and start to fix functions that violate the rules (by changing their behavior, or perhaps in some cases, by changing their documentation to explicitly point out the deviation).
> 
> Also, once again, I am fairly certain this problem is not unique to Oscar within the Julia realm, or in general.

In general there are only few rules possible - to still allow efficient
code.
 - no "!" in the name -> function is safe
 - a list of functions that is guaranteed to return mutable objects (ie.
   single onwer ones)
 - function staring with _, internal, ... -> do only use if you
   understand the consequences.
 - function not exported -> think carefully, maybe ask

much more, and it gets complicated...

 - one can do reference counting
 - stack scanning
 - investigate the julia package supposed to be able to deal with this

But as with the modular stuff above: inplace operations are for the
expert only - and possibly overused in places.

**Comment by tthsqe12** [comment]
I have heeded all of the discussions here and have a much improved more useful version in the works.

File: docs/src/DeveloperDocumentation/new_developers.md (hunk 2 above)
Comment (tthsqe12) [reply docs/src/DeveloperDocumentation/new_developers.md]: I am defending this behaviour as a viable possible interface and have used it in some of my own functions. If it ever is documented, we could name it the 'dangling reference' interface.

File: docs/src/DeveloperDocumentation/new_developers.md (hunk 3 above)
Comment (tthsqe12) [reply docs/src/DeveloperDocumentation/new_developers.md]: The new version might be slightly more helpful in this regard. I point out that ownership conventions are an implementation detail with some small guarantees. The default is therefore that you cannot expect to own anything after passing it around.

**Comment by lkastner** [comment]
The page `new_developers.md` is getting very messy at this point. It looks like there is just being added to it without taking the existing content into consideration. For example now the warnings and overview I originally wrote are in the middle. The "new" opening paragraph "Oscar - what is the idea" seems not to be directed at new developers in particular? Other paragraphs also seem out of place. I would much more recommend having new pages for the new content and just place links to these with a short explanation what these will be about, like already done for the styleguide and building the docs. I would prefer if this page just gave an overview.

File: docs/src/DeveloperDocumentation/new_developers.md (hunk 1 above)
Comment (tthsqe12) [reply docs/src/DeveloperDocumentation/new_developers.md]: I changed Oscar->Julia

**Review by lkastner — APPROVED** [review]

**Review by thofma — APPROVED** [review]

File: docs/src/DeveloperDocumentation/new_developers.md (hunk 2 above)
Comment (fingolfin) [reply docs/src/DeveloperDocumentation/new_developers.md]: Thanks for the explanations. That seems fair. Perhaps we should mark functions doing such things, e.g. by inserting `unsafe` into their name somewhere, or adding `!` or so? (This is not a change request, just wondering)

File: docs/src/DeveloperDocumentation/new_developers.md (hunk 3 above)
Comment (fingolfin) [reply docs/src/DeveloperDocumentation/new_developers.md]: Thank you

File: docs/src/DeveloperDocumentation/debugging.md [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+```
+
+In this example, the modification of `v` is unexpected and may in fact corrupt
+the internal data structures used by the polynomial ring. As such, this
+modification of `v` has to be considered illegal. Upon creation of the array
+called `v`, we have full rights over the object and can mutate at will.
+However, after passing it to the function `PolynomialRing`, we have given up
+*owernership* of the array and are no longer free to modify it.
+
+General Oscar Principle (GOP):
```
Comment (fingolfin) [inline docs/src/DeveloperDocumentation/debugging.md:36]: I am not sure I like being associated with the GOP, but I guess I must read this non-political 😂

**Review by fingolfin — APPROVED** [review]

**Pull request merged**